### PR TITLE
Pygrb pp rewrite

### PR DESCRIPTION
**Pull request by ginoContestabile**
work in progress to replace pylal_cbc_cohptf_sbv_plotter with pycbc executables.

### CODE REVIEW

**Comment by duncanmmacleod** [comment]
What is the plan to rebase this to work with the output of @idorrington92's new `pycbc-multi-inspiral` (HDF5) code and the other HDF5 post-processing scripts I have posted in #2767?

cc @pannarale, @ginoContestabile, @spxiwh, @tcarver1234

**Comment by spxiwh** [comment]
I see a lot of very similar executables here, e.g. pycbc_pygrb_plot_coh_bankveto, pycbc_pygrb_plot_coh_chisquared, .... Can these be made a single executable that just takes the variable that needs to be plotted as an input argument?

**Comment by pannarale** [comment]
Yes, but I think it will be easier to downsize the executables once we 1) the workflow is settled along with what options we (don't) want to have in there, 2) we decide what plots we no longer want.  Right now we are having this pop out all the plots exaclty as they were before.

**Comment by pannarale** [comment]
@idorrington92, @duncanmmacleod, I'm starting to move "xml stuff" in a single place (this includes the glue import spotted by @duncanmmacleod).  This way Tessa and Cameron will have to change things in a single place.

**Comment by spxiwh** [comment]
I'm still concerned that `bin/pygrb/pycbc_pygrb_plot_coh_bankveto` and `bin/pygrb/pycbc_pygrb_plot_coh_autoveto` are basically identical pieces of code, up to one or two lines.  Why not combine these (and a number of the other codes) into a `bin/pygrb/pycbc_pygrb_plot_coh_quantities` and then if it takes `--variable bankveto` on the command line it does what the first code does and `--variable autoveto` does what the second code does. I understand that with XML -> HDF these codes might becomes shorter, but I still think it makes sense for these to be just one executable.

**Comment by pannarale** [comment]
> I'm still concerned that `bin/pygrb/pycbc_pygrb_plot_coh_bankveto` and `bin/pygrb/pycbc_pygrb_plot_coh_autoveto` are basically identical pieces of code, up to one or two lines. Why not combine these (and a number of the other codes) into a `bin/pygrb/pycbc_pygrb_plot_coh_quantities` and then if it takes `--variable bankveto` on the command line it does what the first code does and `--variable autoveto` does what the second code does. I understand that with XML -> HDF these codes might becomes shorter, but I still think it makes sense for these to be just one executable.

I took care of this by coalescing ;-) the scripts by plot family.  One script can handle all the time series plots, another one can handle all the vetoes vs SNRs, a third one does null stats and coincident SNR vs Coherent SNR, and a fourth one deos signle IFO vs coherent SNR.  I prefered to keep 3 and 4 separate because the plots look quite differente.  Otherwise we run the risk of going back to a fancier version of the coh_PTF plotting script but with a bunch of if-elif-else's (which we don't want to do).  I improved the workflow generator along the way.

Hopefull this will help with debugging the travis issues as well.

**Comment by pannarale** [comment]
By the way, right now the workflow generator works with a single trigger file and injeciton file in input.  I would like to wait for the xml --> hdf5 transition before generalizing that.

File: pycbc/results/pygrb_plotting_utils.py [hunk 1]
```diff
@@ -0,0 +1,640 @@
...
+import numpy
+from pycbc.results import save_fig_with_metadata
+# TODO: imports to fix/remove
+try:
+    from glue import segments
+    from glue.ligolw import utils, lsctables, ligolw, table
+except ImportError:
+    MSG = "Fix when tests on top of the O2 PyGRB installation are finished."
+    logging.warning(MSG)
+    sys.exit()
```
Comment (spxiwh) [inline pycbc/results/pygrb_plotting_utils.py]: Remove the `sys.exit` here. I think that's the cause of at least 2 of the 3 travis failures. This should just raise a warning, then let the code fail normally if any of the routines using any of this code gets called.

File: pycbc/results/pygrb_plotting_utils.py [hunk 2]
```diff
@@ -0,0 +1,640 @@
...
+try:
+    from pylal import MultiInspiralUtils
+    from pylal.coh_PTF_pyutils import get_bestnr, get_det_response
+    from pylal.coh_PTF_pyutils import readSegFiles
+    from pylal.dq import dqSegmentUtils
+except ImportError:
+    MSG = "Running the PyGRB post-processing requires a pylal installation"
+    logging.warning(MSG)
+    sys.exit()
```
Comment (spxiwh) [inline pycbc/results/pygrb_plotting_utils.py]: Same here

File: pycbc/results/pygrb_plotting_utils.py [hunk 3]
```diff
@@ -0,0 +1,640 @@
...
+        inj_data.time = [t-grb_time for t in inj_data.time]
+
+    return grb_time, start, end, trig_data, inj_data
+
+
+# =============================================================================
+# Extract trigger/injection data produced by PyGRB
+# =============================================================================
+
+class PygrbFilterOutput(object):
```
Comment (spxiwh) [inline pycbc/results/pygrb_plotting_utils.py]: One thing to consider here (not for this patch!) is to use some of python's class functionality to only load data when needed. For e.g. you don't need to read/calculate `rel_amp_1` in most cases. Using something like https://www.geeksforgeeks.org/getter-and-setter-in-python/, which we use also in the all-sky code, you can only calculate these things once and store it in memory if needed after the first time.

Comment (spxiwh) [reply pycbc/results/pygrb_plotting_utils.py]: ... This would need the HDF input files anyway

**Comment by spxiwh** [comment]
@pannarale See my comment about where I think at least 2 of the Travis failures are coming from. I'm happy that this is in a state where it is ready to merge once these problems can be fixed.

Another thing to consider: As this is now > 200 commits, I would merge this with "squash and merge". However, doing that would no longer recognize that both you and @ginoContestabile worked on this. If you both want to keep correct credit for this if you collapse this down to no more than ~ 5 commits from both of you, it can be merged with "rebase and merge" which will preserve authorship.

**Comment by pannarale** [comment]
Thanks, @spxiwh!  That solved 2 of the 3 build problems.  The first build is the only one failing now: https://travis-ci.org/gwastro/pycbc/jobs/556139038#L2784

**Comment by pannarale** [comment]
I've seen things like [https://github.com/gwastro/pycbc/blob/master/pycbc/results/followup.py#L28](this) around, so I will give that a try when the cluster with my installation is back online.

**Comment by pannarale** [comment]
https://travis-ci.org/gwastro/pycbc/jobs/556310810#L2770 but I'm sure others import matplotlib.pyplot.

**Comment by pannarale** [comment]
Solved.  I had moved the agg-backend stuff down because of codeclimate complaints, but this was causing a failure in the 1st travis test.

Will run my usual full test again now.

At the end of that you (@spxiwh) want me to git rebase -i master down to 5 commits, correct?

**Comment by spxiwh** [comment]
Yes please! Just squash successive commits from you and successive commits from @ginoContestabile into single commits (don't squash commits from you with commits from @ginoContestabile or vice-versa, or you lose authorship history).

**Comment by pannarale** [comment]
Ta-dah!

**Pull request merged**